src/sdk/pynni/nni/pbt_tuner/pbt_tuner.py

```diff
@@ -34,19 +38,104 @@ def exploit_and_explore(bot_trial_info, top_trial_info, factors, epoch, search_s
+            if random.random() < resample_probability:
+                hyper_parameters[key] = parameter_expressions.loguniform(lb, ub, random_state)
+            elif random.random() > 0.5:
+                hyper_parameters[key] = min(np.exp(min(np.log(hyper_parameters[key]) + perturb, np.log(ub))), ub)
+            else:
+                hyper_parameters[key] = max(np.exp(max(np.log(hyper_parameters[key]) - perturb, np.log(lb))), lb)
+        elif search_space[key]["_type"] == "qloguniform":
+            lb, ub, q = search_space[key]["_value"][:3]
+            multi = hyper_parameters[key] // q
+            if random.random() < resample_probability:
+                hyper_parameters[key] = parameter_expressions.qloguniform(lb, ub, q, random_state)
+            elif random.random() > 0.5:
+                hyper_parameters[key] = min((multi + 1) * q, ub)
+            else:
+                hyper_parameters[key] = max((multi - 1) * q, lb)
+        elif search_space[key]["_type"] == "normal":
+            mu, sigma = search_space[key]["_value"][:2]
+            perturb = sigma * factor
+            if random.random() < resample_probability:
+                hyper_parameters[key] = parameter_expressions.normal(mu, sigma, random_state)
+            elif random.random() > 0.5:
+                hyper_parameters[key] = hyper_parameters[key] + perturb
+            else:
+                hyper_parameters[key] = hyper_parameters[key] - perturb
+        elif search_space[key]["_type"] == "qnormal":
+            mu, sigma, q = search_space[key]["_value"][:3]
+            if random.random() < resample_probability:
+                hyper_parameters[key] = parameter_expressions.qnormal(mu, sigma, q, random_state)
+            elif random.random() > 0.5:
+                hyper_parameters[key] = hyper_parameters[key] + q
+            else:
+                hyper_parameters[key] = hyper_parameters[key] - q
+        elif search_space[key]["_type"] == "lognormal":
+            mu, sigma = search_space[key]["_value"][:2]
+            perturb = sigma * factor
+            if random.random() < resample_probability:
+                hyper_parameters[key] = parameter_expressions.lognormal(mu, sigma, random_state)
+            elif random.random() > 0.5:
+                hyper_parameters[key] = np.exp(np.log(hyper_parameters[key]) + perturb)
+            else:
+                hyper_parameters[key] = np.exp(np.log(hyper_parameters[key]) - perturb)
+        elif search_space[key]["_type"] == "qlognormal":
+            mu, sigma, q = search_space[key]["_value"][:3]
+            if random.random() < resample_probability:
+                hyper_parameters[key] = parameter_expressions.qlognormal(mu, sigma, q, random_state)
+            elif random.random() > 0.5:
+                hyper_parameters[key] = hyper_parameters[key] + q
+            else:
+                hyper_parameters[key] = max(0, hyper_parameters[key] - q)
```

Comment:
suggest using eps like 1e-10 instead of 0

Reply:
modified